homework/index.js

```diff
@@ -30,17 +38,220 @@
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
-      }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+  function renderError(error) {
+    createAndAppend('div', root, {
+      text: error.message,
+      class: 'alert alert-danger display-3 text-center',
     });
   }
 
-  const HYF_REPOS_URL = 'https://api.github.com/orgs/HackYourFuture/repos?per_page=100';
+  function contributorStats(contribution, cardBody, dataArr, textValue) {
+    const parent = createAndAppend('a', cardBody, {
+      class: 'card-text btn btn-danger d-flex justify-content-between rounded-0 my-3',
+      text: `${textValue}:`,
+      href: `https://github.com/${contribution.login}?tab=${textValue}`,
+      target: '_blank',
+    });
+    createAndAppend('span', parent, {
+      class: 'card-text badge badge-light d-flex justify-content-between ',
+      text: dataArr.length,
+    });
+  }
+
+  function fetchStats(contribution, cardBody, followingsUrl) {
```

Comment:
Here is a solution that takes advantage of `Promise.all()`, allowing the browser to do up to 6 concurrent requests as we demoed in class:

```js
function fetchStats(contribution, cardBody, followingsUrl) {
  const promises = [
    contribution.followers_url,
    followingsUrl,
    contribution.subscriptions_url,
  ].map(url => fetchJSON(url));

  Promise.all(promises)
    .then(results => {
      ['followers', 'following', 'subscriptions'].forEach((category, index) => {
        contributorStats(contribution, cardBody, results[index], category);
      });
    })
    .catch(err => renderError(err));
}
```

Reply:
Thank you Jim, I am going to implement it asap.

Reply:
I want to say something about the difference between chaining the promises and the promise all. In Promise.all the response waits till all the promises resolve. I have tested I have given wrong link to one of the request in Promise.all the the requests haven't been resolved even 2 of them have to be resolved, this shows if one request is rejected the rest is going to be rejected. On the other hand in promise chaining it continues till the response is rejected as an error. I think using promise all vs promise chain differs in case of scenario.

Reply:
That’s a correct assessment. If not all promises need to be resolved for the result to be successful then you should not use Promise.all().

Reply:
Also I want to ask, in the code above we used map() to get array of promises, I removed map and tested with array of urls. It is still working in Promise.all(). I'm in shock right now. Did i miss something, because when I remove map i also remove fetchJSON() inside the map. How is it going to make request without fetchJSON()?

Reply:
``  function fetchStats(contribution, cardBody, followingsUrl) {
    const promises = [contribution.followers_url, followingsUrl, contribution.subscriptions_url];

    Promise.all(promises)
      .then(results => {
        ['followers', 'following', 'subscriptions'].forEach((category, index) => {
          contributorStats(contribution, cardBody, results[index], category);
        });
      })
      .catch(err => renderError(err));
  }

Reply:
`function fetchStats(contribution, cardBody, followingsUrl) {
    const promises = [contribution.followers_url, followingsUrl, contribution.subscriptions_url];
  Promise.all(promises)
      .then(results => {
        ['followers', 'following', 'subscriptions'].forEach((category, index) => {
          contributorStats(contribution, cardBody, results[index], category);
        });
      })
      .catch(err => renderError(err));
  }`